File: solutions/0322_Coin_Change/dp.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        for (int i = 1; i <= amount; ++i) {
            dp[i] = -1;
            int best = Integer.MAX_VALUE;
            for (int c : coins) {
                if (i >= c && dp[i - c] >= 0) {
                    best = Math.min(best, dp[i - c] + 1);
                }
            }
            if (best < Integer.MAX_VALUE) {
                dp[i] = best;
            }
        }
        return dp[dp.length - 1];
    }
}

File: solutions/0322_Coin_Change/dp.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [1 << 31] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            best = 1 << 31
            for c in coins:
                if i >= c:
                    best = min(best, dp[i - c] + 1)
            dp[i] = best
        return -1 if dp[amount] == 1 << 31 else dp[amount]
